src/kino_telegram.nim

```diff
@@ -0,0 +1,169 @@
+import std/[asyncdispatch, options, json, strutils, strformat, tables]
+import ws, telebot
+import questionable
+import ./protocol
+import telegram/config
+
+type
+  Client = object
+    user: User
+    ws: WebSocket
+  
+  Server = object
+    clients: Table[int, Client]
+    host: string
+    playlist: seq[string]
+
+let cfg = getConfig()
+  
+var
+  server: Server
+  bot: Telebot
+
+proc getClient(server: Server, user: User): ?Client =
+  if server.clients.hasKey(user.id):
+    return some server.clients[user.id]
+  
+proc send(client: ?Client, data: Event) =
+  if fut =? client.?ws.?send($(%data)):
+    asyncCheck fut
+
+proc showEvent(client: Client, text: string) {.async.} =
+  discard await bot.sendMessage(client.user.id, text)
+
+proc join(client: Client): Future[bool] {.async.} =
+  echo "Joining.."
+  await client.ws.send($(%Auth(client.user.username.get, "")))
+
+  let resp = unpack(await client.ws.receiveStrPacket())
+  match resp:
+    Joined(newName, newRole):
+      await client.showEvent("Welcome to kinoplex!")
+    Error(reason):
+      await client.showEvent("bridge failed to join: " & reason)
+      result = true
+    _: discard
+
+proc handleServer(client: Client) {.async.} =
+  if await client.join(): return
+  client.ws.setupPings(5)
+  while client.ws.readyState == Open:
+    let event = unpack(await client.ws.receiveStrPacket())
+    match event:
+      Message(name, text):
+        discard await bot.sendMessage(client.user.id, name & ": " & text)
+      Clients(names):
+        await client.showEvent("Users: " & names.join(", "))
+      Joined(name, role):
+        await client.showEvent(&"{name} joined as {role}")
+      Left(name):
+        await client.showEvent(name & " left")
+      Jannies(jannies):
+        if jannies.len < 1:
+          await client.showEvent("There are currently no jannies")
+        else:
+          await client.showEvent("Jannies: " & jannies.join(", "))
+      PlaylistAdd(url):
+        server.playlist.add url
+      PlaylistPlay(index):
+        await client.showEvent("Playing " & server.playlist[index])
+      PlaylistClear:
+        server.playlist.setLen(0)
+        await client.showEvent("Playlist cleared")
+      Error(reason):
+        await client.showEvent(reason)
+      Null: discard
+      Auth: discard
+      Success: discard
+      _: discard
+                  
+  close client.ws
+  server.clients.del(client.user.id)
+
+proc validUrl(url: string; acceptFile=false): bool =
+  url.len > 0 and "\n" notin url and (acceptFile or "http" in url)
+
+template handlerizerKinoplex(body: untyped): untyped =
+  proc cb(bot: Telebot, c: Command): Future[bool] {.gcsafe async.} =
+    if c.message.fromUser.isNone: return
+
+    let
+      client {.inject.} = server.getClient(!c.message.fromUser)
+      message {.inject.} = c.message
+    body
```

Comment:
`if client.isNone: return`
Any reason not to do this here before running body? It's done on line 107 and it seems like it should be done in every case where this handlerizer could be used.

Reply:
I initially avoided it because it would result in a lot of `client.get` being redundant, but now that's not the case anymore

Reply:
I have refactored the handlerizer and usage of `Option[Client]`